_posts/2023-01-07-datafusion-16.0.0.md

```diff
@@ -0,0 +1,289 @@
+such as
+
+- Cost based optimizer (CBO) that can automatically reorder join evaluations, algorithm (Merge / Hash), and pick build side based on available statistics join type (`INNER`, `LEFT`, etc) (#4219)
+- Fast non `column=column` equijoins such as `JOIN ON a.x + 5 = b.y`
+- Better optimized non-equijoins (nested loops rather than cross join and filter) (#4562) <!-- TODO is this a good thing to mention as any time this is usd the query is going to go slow or the data size is small -->
+
+# Streaming Execution
+
+TODO: write up some supporting text introducing this concept and explaining that DataFusion can now be used for stream processing, much like [Spark Streaming](https://www.databricks.com/glossary/what-is-spark-streaming)
+
+Support for executing infinite files and boundedness-aware join reordering rule (#4694)
+
+# Better Support for Distributed catalogs
+
+With the rise of distributed catalog / metadata stores such as
+delta.io and iceberg (TODO get links to rust projects), it is
+increasingly important to be enable asynchronous I/O to those remote
+catalogs during query planning rather than requiring up front fast
+local access to all relevant catalog information.
+
+
+DataFusion has been enhanced with better support for such asynchronous catalogs (#4607) , which also resulted in improved configuration management APIs.
+
+
+
+# Additional SQL support
+SQL support continues to improve, including some of these highlights:
+
+- Add TPC-DS query planning regression tests (#4719)
+- feat: support prepare statement (#4490)
+- Implement cast between Date and Timestamp (#4726)
+- Support type coercion for timestamp and utf8 (#4312)
+- Full support for time32 and time64 literal values (`ScalarValue`) (#4156)
+- add uuid() function to return a unique uuid per row (#4041)
+- Implement `current_time` scalar function (#4054)
+- Implement current_date scalar function (#4022)
+- Compressed CSV/JSON support (#3642)
+
+The community has also been investing in sqllogic based tests to help keep DataFusion's quality high with less work (TODO add some more detail / lnks)
+
+
+# Substrait
+
+TODO motivating introduction of substrait and why this is interesting
+
+We moved physical plan serde from Ballista to DataFusion (#4390) and we are in the proceess of adding support substrait
+
+
+<!--
+Andrew Lamb: I tried to work in a mention of the python bindings and encouraging a champion for them to step forward, however it felt more like it should be a separate post :thinking:
+-->
+
+## Looking for help:
+
+DataFusion is currently targeted at developers building other systems, not at end users (e.g. data scientists) themselves. Given the currently growing number of systems built using DataFusion, it seems well suited for this purpose.
+
+DataFusion has basic python bindings which has the potential to expand datafusion  to more end users a major missing piece are the python bindings
+
+
+# python bindings and growing the community and ecosystem
```

Comment:
I tried to work in a mention of the python bindings and encouraging a champion for them to step forward, however it felt more like it should be a separate post 🤔  -- @andygrove  what do you think about a post describing the python bindings, why they are cool, and trying to find people to help drive that project?

Reply:
I can't figure out how to work this in so I think we should write another post -- maybe on a different site

I took the content / notes I had and put them in a google doc: https://docs.google.com/document/d/1zNfK8pIOqgHURX2lHK0JhSKTaCH3t23tYDKpGLWdFRY/edit

Reply:
Thanks. I agree. I can work on the Python post